cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactRepository.java

```diff
@@ -242,6 +253,109 @@ public ArtifactDetail addArtifact(Id.Artifact artifactId, File artifactFile,
     }
   }
 
+  /**
+   * Scan all files in the local system artifact directory, looking for jar files and adding them as system artifacts.
+   * If the artifact already exists it will not be added again unless it is a snapshot version.
+   *
+   * @throws IOException if there was some IO error adding the system artifacts
+   * @throws WriteConflictException if there was a write conflicting adding the system artifact. This shouldn't happen,
+   *                                but if it does, it should be ok to retry the operation.
+   */
+  public void addSystemArtifacts() throws IOException, WriteConflictException {
+
+    // scan the directory for artifact .jar files and config files for those artifacts
+    List<SystemArtifactConfig> systemArtifacts = new ArrayList<>();
+    for (File jarFile : DirUtils.listFiles(systemArtifactDir, "jar")) {
+      // parse id from filename
+      Id.Artifact artifactId;
+      try {
+        artifactId = parse(Id.Namespace.SYSTEM, jarFile.getName());
+      } catch (InvalidArtifactException e) {
+        LOG.warn(String.format("Skipping system artifact '%s' because the name is invalid: ", e.getMessage()));
+        continue;
+      }
+
+      // check for a corresponding .json config file
+      String artifactFileName = jarFile.getName();
+      String configFileName = artifactFileName.substring(0, artifactFileName.length() - ".jar".length()) + ".json";
+      File configFile = new File(systemArtifactDir, configFileName);
+      try {
+        if (configFile.isFile()) {
+          // if a config file exists, parse it and add it to the list
+          systemArtifacts.add(SystemArtifactConfig.read(artifactId, configFile, jarFile));
+        } else {
+          // otherwise, don't parse it
+          systemArtifacts.add(SystemArtifactConfig.builder(artifactId, jarFile).build());
+        }
+      } catch (InvalidArtifactException e) {
+        LOG.warn(String.format("Could not add system artifact '%s' because it is invalid.", artifactFileName), e);
+      }
+    }
+
+
+    // Need to be sure to add parent artifacts before artifacts that extend them.
```

Comment:
How are making sure that we add parent artifacts before adding artifacts that extend them? Does Collections.sort do the trick?